src/main/java/com/bio4j/angulillos/TypedLinkGraph.java

```diff
@@ -0,0 +1,128 @@
+package com.bio4j.angulillos;
+
+public abstract class TypedLinkGraph<
+  G extends TypedLinkGraph<G,RV,RE>,
+  RV,RE
+>
+extends TypedGraph<G,RV,RE> {
+
+  protected TypedLinkGraph(UntypedGraph<RV,RE> raw) { super(raw); }
+
+  public abstract class LinkEdge<
+    SG extends TypedGraph<SG,RV,RE>,
+    S  extends TypedGraph<SG,RV,RE>.Vertex<S>,
+    E  extends LinkEdge<SG,S, E, TG,T>,
+    TG extends TypedGraph<TG,RV,RE>,
+    T  extends TypedGraph<TG,RV,RE>.Vertex<T>
+  > extends Element<E, LinkEdgeType<SG,S, E, TG,T>, RE>
+    implements TypedEdge<
+      S, TypedGraph<SG,RV,RE>.VertexType<S>,
+      E, LinkEdgeType<SG,S, E, TG,T>,
+      T, TypedGraph<TG,RV,RE>.VertexType<T>,
+      G,RV,RE
+    > {
+      protected LinkEdge(RE raw, LinkEdgeType<SG,S, E, TG,T> type) { super(raw, type); }
+    }
+
+  public abstract class LinkEdgeType<
+    SG extends TypedGraph<SG,RV,RE>,
+    S  extends TypedGraph<SG,RV,RE>.Vertex<S>,
+    E  extends LinkEdge<SG,S, E, TG,T>,
+    TG extends TypedGraph<TG,RV,RE>,
+    T  extends TypedGraph<TG,RV,RE>.Vertex<T>
+  > extends ElementType<E, LinkEdgeType<SG,S, E, TG,T>, RE>
+    implements TypedEdge.Type<
+      S, TypedGraph<SG,RV,RE>.VertexType<S>,
+      E, LinkEdgeType<SG,S, E, TG,T>,
+      T, TypedGraph<TG,RV,RE>.VertexType<T>,
+      G,RV,RE
+    > {
+
+      protected LinkEdgeType<SG,S, E, TG,T> self() { return this; }
+
+      private final TypedGraph<SG,RV,RE>.VertexType<S> sourceType;
+      private final TypedGraph<TG,RV,RE>.VertexType<T> targetType;
+
+      @Override public final TypedGraph<SG,RV,RE>.VertexType<S> sourceType() { return this.sourceType; }
+      @Override public final TypedGraph<TG,RV,RE>.VertexType<T> targetType() { return this.targetType; }
+
+      // NOTE: this initializer block will be inherited and will add each edge type to the set
+      {
+        if (
+          TypedLinkGraph.this.edgeTypes.removeIf( (AnyEdgeType et) ->
+            et._label().equals( self()._label() )
+          )
+        ) {
+          throw new IllegalArgumentException("The graph contains duplicate edge type: " + self()._label());
+        }
+
+        TypedLinkGraph.this.edgeTypes.add(self());
+      }
+
+      protected LinkEdgeType(
+        TypedGraph<SG,RV,RE>.VertexType<S> sourceType,
+        TypedGraph<TG,RV,RE>.VertexType<T> targetType
+      ) {
+
+        this.sourceType = sourceType;
+        this.targetType = targetType;
+
+        sourceType.outEdges.add( self() );
```

Comment:
Similarly to the sets of "untyped" edges and vertices in the `TypedGraph`, each [vertex type](https://github.com/bio4j/angulillos/blob/master/src/main/java/com/bio4j/angulillos/AnyVertexType.java#L12-L13) has a set of the in/out edge types. It may be needed for the automatic schema creation (or depending on the backend it could be used for some advanced vertex type tuning).
Also I think we discussed this in the context of #63 for schema inspection.